orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,743 @@
+#include "pfcwdorch.h"
+#include "saiserialize.h"
+#include "portsorch.h"
+#include "converter.h"
+#include "redisapi.h"
+
+#define PFC_WD_ACTION                   "action"
+#define PFC_WD_DETECTION_TIME           "detection_time"
+#define PFC_WD_RESTORATION_TIME         "restoration_time"
+
+#define PFC_WD_QUEUE_STATUS             "status"
+#define PFC_WD_QUEUE_STATUS_OPERATIONAL "operational"
+#define PFC_WD_QUEUE_STATUS_STORMED     "stormed"
+
+#define PFC_WD_DETECTION_TIME_MAX       (5 * 1000)
+#define PFC_WD_DETECTION_TIME_MIN       100
+#define PFC_WD_RESTORATION_TIME_MAX     (60 * 1000)
+#define PFC_WD_RESTORATION_TIME_MIN     100
+#define PFC_WD_TC_MAX                   8
+
+extern sai_port_api_t *sai_port_api;
+extern sai_queue_api_t *sai_queue_api;
+extern PortsOrch *gPortsOrch;
+
+PfcWdOrch::PfcWdOrch(DBConnector *db, vector<string> &tableNames):
+    Orch(db, tableNames)
+{
+    SWSS_LOG_ENTER();
+}
+
+PfcWdOrch::~PfcWdOrch(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+void PfcWdOrch::doTask(Consumer& consumer)
+{
+    SWSS_LOG_ENTER();
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string key = kfvKey(t);
+        string op = kfvOp(t);
+
+        if (op == SET_COMMAND)
+        {
+            createEntry(key, kfvFieldsValues(t));
+        }
+        else if (op == DEL_COMMAND)
+        {
+            deleteEntry(key);
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+        }
+
+        consumer.m_toSync.erase(it++);
+    }
+}
+
+PfcWdOrch::PfcWdAction PfcWdOrch::deserializeAction(const string& key)
+{
+    SWSS_LOG_ENTER();
+
+    const std::map<std::string, PfcWdAction> actionMap =
+    {
+        { "forward", PfcWdAction::PFC_WD_ACTION_FORWARD },
+        { "drop", PfcWdAction::PFC_WD_ACTION_DROP },
+    };
+
+    if (actionMap.find(key) == actionMap.end())
+    {
+        return PfcWdAction::PFC_WD_ACTION_UNKNOWN;
+    }
+
+    return actionMap.at(key);
+}
+
+void PfcWdOrch::createEntry(const string& key, const vector<FieldValueTuple>& data)
+{
+    SWSS_LOG_ENTER();
+
+    uint32_t detectionTime = 0;
+    uint32_t restorationTime = 0;
+    PfcWdAction action = PfcWdAction::PFC_WD_ACTION_UNKNOWN;
```

Comment:
action can be set to drop by default and the validation logic below could be removed.

Reply:
done